src/main/java/hudson/tasks/test/AbstractTestResultAction.java

```diff
@@ -355,19 +623,612 @@ private CategoryDataset buildDataSet(StaplerRequest req) {
+                    toolTipString += ",...";
+                    continue;
+                }
+                if (!toolTipString.equals(""))
+                    toolTipString += ", ";
+                toolTipString += caseName;
+            }
+            NumberOnlyBuildLabel label = new NumberOnlyBuildLabel(testResultAction.run);
+            dataset.add(lengthyTestCount, "Lengthy Tests", label);
+
+            /*
+             * Also being stored in {@link #lengthyToolTip} in order to generate tooltips on hovering mouse
+             * over the trend.
+             */
+            lengthyToolTip.put(label, toolTipString);
+            stack.pop();
+        }
+        LOGGER.log(Level.FINER, "total test trend count for {0}: {1}", new Object[]{run, count});
+        return dataset.build();
+    }
+
+
+    /**
+     * Utility method to count number of test flappers in the the build present at the head of buildHistory.
+     * @param buildHistory Queue to represent the past builds considered to check whether a testcase flapped in any
+     *                     of the past builds. If it flapped then it will be counted as flapper.
+     * @param testsHistory Map representing the builds in which a testcase flapped for each testcase by storing a mapping
+     *                     from testcase to the queue of builds in which it flapped.
+     */
+    private void shiftBuildHistoryUtil (ArrayDeque<Pair<AbstractTestResultAction<?>, HashSet<Integer>>> buildHistory,
+                                        Map<Integer, ArrayDeque<AbstractTestResultAction<?>>> testsHistory) {
+        Pair<AbstractTestResultAction<?>, HashSet<Integer>> lastBuild = buildHistory.peek();
+        buildHistory.remove();
+        AbstractTestResultAction<?> a = lastBuild.first;
+        HashSet<Integer> buildSet = lastBuild.second;
+        int count = 0;
+        for (Integer testCaseIndex : buildSet) {
+            ArrayDeque<AbstractTestResultAction<?>> testHistory = testsHistory.get(testCaseIndex);
+            while (testHistory != null && !testHistory.isEmpty() && testHistory.peek().run.number >= a.run.number) {
+                testHistory.remove();
+            }
+            if (testHistory != null && !testHistory.isEmpty()) {
+                count++;
+            }
+        }
+        flapperCountToolTip.put(a.run.number, count);
+    }
+
+    /**
+     * A method to build dataset for the chosen project in order to generate trends for the analysis of
+     * test flappers and display builds at which they failed for selected testcases.
+     * @param req The HTTP request message for the particular project or all projects for "flaky tests"
+     *            trend type.
+     * @return An object of type {@link XYDataset} which contains each testcase as separate data series and each
+     * data series contains (x,y) points depicting builds at which that particular testcase failed.
+     */
+    private XYDataset buildFlapperDataset (StaplerRequest req) {
```

Comment:
Avoid giving unnecessary whitespace b/w two lines inside a method. Please do same for all the changes.

Reply:
Done.